File: typegate/tests/runtimes/wasm_wire/wasm_duplicate.ts
// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

import { Policy, t, typegraph } from "@typegraph/sdk/index.ts";
import { WasmRuntime } from "@typegraph/sdk/runtimes/wasm.ts";

export const tg = await typegraph("wasm-duplicate-ts", (g: any) => {
  const wasm1 = WasmRuntime.wire("rust.wasm");
  const wasm2 = WasmRuntime.wire("rust.wasm");

  g.expose(
    {
      add1: wasm1
        .handler(
          t.struct({ a: t.float(), b: t.float() }).rename("add_args"),
          t.integer(),
          { name: "add" },
        )
        .rename("add"),
      add2: wasm2
        .handler(
          t.struct({ a: t.float(), b: t.float() }).rename("add_args2"),
          t.integer(),
          { name: "add" },
        )
        .rename("add2"),
    },
    Policy.public(),
  );
});

File: tests/runtimes/wasm_reflected/wasm_duplicate.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from typegraph.graph.typegraph import Graph
from typegraph.policy import Policy
from typegraph.runtimes.wasm import WasmRuntime

from typegraph import t, typegraph


@typegraph()
def wasm_duplicate(g: Graph):
    pub = Policy.public()
    wasm1 = WasmRuntime.reflected("rust.wasm")
    wasm2 = WasmRuntime.reflected("rust.wasm")

    g.expose(
        test1=wasm1.export(
            t.struct({"a": t.float(), "b": t.float()}),
            t.integer(),
            name="add",
        ).with_policy(pub),
        test2=wasm2.export(
            t.struct({"a": t.float(), "b": t.float()}),
            t.integer(),
            name="add",
        ).with_policy(pub),
    )
